zsh-history-substring-search.zsh

```diff
@@ -233,20 +234,33 @@ _history-substring-search-begin() {
     _history_substring_search_query=$BUFFER
 
     #
-    # $BUFFER contains the text that is in the command-line currently.
-    # we put an extra "\\" before meta characters such as "\(" and "\)",
-    # so that they become "\\\(" and "\\\)".
+    # compose search pattern
     #
-    local escaped_query=${BUFFER//(#m)[\][()|\\*?#<>~^]/\\$MATCH}
+    # (q)        - escape special characters within string
+    # (s:CHAR>:) - split string by CHAR
+    # j:CHAR:)   - join array to string with CHAR as seperator
+    # =          - split string in arguments
 
+    if [ "$HISTORY_SUBSTRING_SEARCH_METHOD" = 'fuzzy_words' ]; then
+      # find history by words in given order
+      _history_substring_search_query_parts=(${=_history_substring_search_query})
+    elif [ "$HISTORY_SUBSTRING_SEARCH_METHOD" = 'fuzzy_chars' ]; then
+      # find history by chars in given order
+      _history_substring_search_query_parts=(${(s::)_history_substring_search_query})
+    else
+      # find history by query
+      _history_substring_search_query_parts=(${_history_substring_search_query})
+    fi
+    local search_pattern=${(j:*:)${(q)_history_substring_search_query_parts}}
+    
     #
-    # Find all occurrences of the search query in the history file.
+    # Find all occurrences of the search pattern in the history file.
     #
     # (k) returns the "keys" (history index numbers) instead of the values
     # (R) returns values in reverse older, so the index of the youngest
     # matching history entry is at the head of the list.
     #
-    _history_substring_search_raw_matches=(${(k)history[(R)(#$HISTORY_SUBSTRING_SEARCH_GLOBBING_FLAGS)*${escaped_query}*]})
+    _history_substring_search_raw_matches=( $(setopt EXTENDED_GLOB; echo ${(k)history[(R)(#$HISTORY_SUBSTRING_SEARCH_GLOBBING_FLAGS)*${search_pattern}*]}))
```

Comment:
It turned out I didn't need setop extendedglob at all, cause it is already set first line of function